### ISSUE TITLE
reading from InMemoryCache causes memory leak (in React Native)

### ISSUE THREAD

**Issue by DanijelBojcic** [apollographql/apollo-client#8903]
**Intended outcome:**

Calling `readFragment`/`writeFragment` should not increase memory usage at all.

**Actual outcome:**

Callling `readFreagment` and then `writeFragment` repeatedly will increase the memory usage.
Calling only `readFragment` without `writeFragment` will only increase the memory once, which I guess is some caching to avoid unnecessary reads?

**How to reproduce the issue:**

I have created a simple app with reproduction: [apollo-memory-leak](https://github.com/DanijelBojcic/apollo-memory-leak)

**Versions**

```
  System:
    OS: macOS 11.5.2
  Binaries:
    Node: 14.16.0 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.11 - /usr/local/bin/npm
  Browsers:
    Chrome: 94.0.4606.71
    Safari: 15.0
  npmPackages:
    @apollo/client: ^3.4.16 => 3.4.16 
```

**Comment by DanijelBojcic** [comment]
@brainkim React really seems to level out at ~15 MB. 
The issue is originally coming from a big react-native project I'm working on, its a lon running application and its crashing due to memory leak after 1 day. I made the reproduction in react only because it's easier to kickstart. 

Here's a React-Native reproduction app: [RN-apollo-memory-leak](https://github.com/DanijelBojcic/RN-apollo-memory-leak)

When I profile the apps memory usage in Android Studio it starts around 140 MB. After triggering the read/write cycle once its 185 MB, triggering it again and its 205MB. Calling the garage collector will only release around 10 MB.

![Screenshot 2021-10-19 at 06 23 15](https://user-images.githubusercontent.com/50402950/137844710-d1863c78-1aad-4e99-8f56-d253d42de821.png)

### Versions:
```
Android
"@apollo/client": "3.4.16"
"graphql": "15.6.1"
"react": "17.0.2"
"react-native": "0.66.0"
```

**Comment by phryneas** [comment]
We've recently released Apollo Client 3.9 which should solve this problem in most cases. 

In React Native, we cannot completely solve it as they lack support for `WeakRef` and `FinalizationRegistry`, but the problem should be more manageable now due to a lot fixes included in that version. 
We also added [a lot of configuration options](https://www.apollographql.com/docs/react/caching/memory-management/) to allow you to tweak a lot of different cache sizes, so you should be able to optimize your application further now.

As there is nothing more we can do on our side, I'm going to close this issue. As soon as React Native adds the missing capabilities, the code will automatically pick them up and work even better.
